### PR TITLE
[MAINT, MRG] Refactor TimeMixin to include more time operations, ignore filter warning for decim for EpochsTFR

### PR DESCRIPTION
**Pull request by alexrockhill**
Fix leftovers from https://github.com/mne-tools/mne-python/pull/10940.

So I thought, might as well loop in `Evoked` with this to be even more DRY but in doing so it turns out that they actually decimate in different ways. It comes down to the order of the following lines:

Evoked method:
```
self._decim *= decim
start_idx = int(round(-self._raw_times[0] * (self.info['sfreq'] *
                                                     self._decim)))
```

Epochs method:
```
start_idx = int(round(-self._raw_times[0] * (self.info['sfreq'] *
                                                     self._decim)))
self._decim *= decim
```

What this leads to is:

Original times:

```
array([-1. , -0.9, -0.8, -0.7, -0.6, -0.5, -0.4, -0.3, -0.2, -0.1,  0. ,
        0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9])
```

Evoked method:

```
array([-1. , -0.4,  0.2,  0.8])
```

Epochs method:

```
array([-0.6,  0.0,  0.6])
```

cc @agramfort @drammock @larsoner

Seems fairly consequential...

### CODE REVIEW

**Comment by alexrockhill** [comment]
Also there's a bit of an issue that the `mixin`s docstring can only say one return type when self is returned which leads to issues like here https://mne.tools/stable/generated/mne.Evoked.html#mne.Evoked.shift_time where evoked says it returns an Epochs object.

**Comment by alexrockhill** [comment]
After a bit of digging, it seems like it's because the `Epochs` must have 0 because that's the sample that they are defined based on, which seems reasonable, it just would make sense that the `Evoked` would do the same.

**Comment by alexrockhill** [comment]
Ok this should fix all the things I found that were inconsistent.

**Comment by alexrockhill** [comment]
That looks pretty sensible to me for the multiple return type https://output.circle-artifacts.com/output/job/76871b03-098a-4353-9ac0-402f230ccd3b/artifacts/0/dev/generated/mne.Evoked.html?highlight=evoked#mne.Evoked.decimate, definitely an improvement over `main`.

**Comment by alexrockhill** [comment]
I realized I wasn't helping too much by putting another `mixin` that was redundant with `TimeMixin` so I just refactored everything within `TimeMixin` which really trimmed down a lot of redundant code. This got a bit more than I thought but I think this is a really nice refactoring.

**Comment by alexrockhill** [comment]
Ah shoot, times just has to be set with the _set_times setter one more place, I'll fix it tomorrow morning.

**Comment by drammock** [comment]
> That looks pretty sensible to me for the multiple return type https://output.circle-artifacts.com/output/job/76871b03-098a-4353-9ac0-402f230ccd3b/artifacts/0/dev/generated/mne.Evoked.html?highlight=evoked#mne.Evoked.decimate, definitely an improvement over `main`.

I'm not sure about the glossary entry for "time-series object".  Try `git grep "inst :"` and you'll see lots of examples of how we handle this situation elsewhere. We're not perfectly consistent in how we do it, but let's not make that worse by adding a new variant.

**Comment by drammock** [comment]
cross-ref to #10947.  Can you touch those two examples/tutorials in this PR, so they will render here and we can be sure they work again?

File: doc/changes/latest.inc [hunk 1]
```diff
@@ -208,3 +208,5 @@ API and behavior changes
 - The FLASH BEM surface generation via :func:`mne.bem.make_flash_bem` or the :ref:`mne flash_bem` command has now been clarified in terms of expected inputs. In :func:`mne.bem.convert_flash_mris` the ``convert`` parameter is now deprecated and the ``flash5`` and ``flash30`` now accept list of file names or nibabel images. One breaking change is that now the flash files are looked for as mef05_*.mgz and mef30_*.mgz in mri/flash folder (:gh:`10715` by `Alex Gramfort`_)
 
 - The ``to_data_frame`` methods for :meth:`Raw <mne.io.Raw.to_data_frame>`, :meth:`Epochs <mne.Epochs.to_data_frame>`, :meth:`Evoked <mne.Evoked.to_data_frame>`, :meth:`SourceEstimate <mne.SourceEstimate.to_data_frame>`, :meth:`EpochsTFR <mne.time_frequency.EpochsTFR.to_data_frame>`, and :meth:`AverageTFR <mne.time_frequency.AverageTFR.to_data_frame>` all now have default ``time_format=None`` (keep time in seconds) instead of ``time_format='ms'`` (convert time to milliseconds) (:gh:`10925` by `Daniel McCloy`_)
+
+- :func:`mne.Evoked.decimate` now decimates keeping the sample at time 0 when the event occurred to be consistent with :func:`mne.Epochs.decimate` (:gh:`10945` by `Alex Rockhill`_)
```
Comment (drammock) [inline doc/changes/latest.inc:212]:
```suggestion
- :func:`mne.Evoked.decimate` now aligns to the sample at time=0 (so that the exact sample when the event occurred is preserved) to be consistent with :func:`mne.Epochs.decimate` (:gh:`10945` by `Alex Rockhill`_)
```

File: mne/tests/test_evoked.py [hunk 2]
```diff
@@ -81,7 +81,7 @@ def test_decim():
     info = create_info(n_channels, sfreq, 'eeg')
     with info._unlock():
         info['lowpass'] = sfreq_new / float(decim)
-    evoked = EvokedArray(data, info, tmin=-1)
+    evoked = EvokedArray(data, info)
```
Comment (drammock) [inline mne/tests/test_evoked.py:84]: why change this?  I get that alignment will change now, so the exact times will change on the `assert` lines below, but an EvokedArray with negative `tmin` should still work.

File: mne/time_frequency/tests/test_tfr.py [hunk 3]
```diff
@@ -715,7 +715,7 @@ def test_add_channels():
     tfr_badsf = tfr_eeg.copy()
     with tfr_badsf.info._unlock():
         tfr_badsf.info['sfreq'] = 3.1415927
-    tfr_eeg = tfr_eeg.crop(-.1, .1)
+    tfr_eeg = tfr_eeg.crop(.1, .1)
```
Comment (drammock) [inline mne/time_frequency/tests/test_tfr.py:718]: cropping to exactly one sample? why this change?

File: mne/utils/mixin.py [hunk 4]
```diff
@@ -423,8 +423,33 @@ def metadata(self, metadata, verbose=None):
...
+        if decim > 1 and lowpass is None:
+            warn('The measurement information indicates data is not low-pass '
+                 'filtered. The decim=%i parameter will result in a sampling '
+                 'frequency of %g Hz, which can cause aliasing artifacts.'
+                 % (decim, new_sfreq))
+        elif decim > 1 and new_sfreq < 3 * lowpass:
+            warn('The measurement information indicates a low-pass frequency '
+                 'of %g Hz. The decim=%i parameter will result in a sampling '
+                 'frequency of %g Hz, which can cause aliasing artifacts.'
+                 % (lowpass, decim, new_sfreq))  # > 50% nyquist lim
```
Comment (drammock) [inline mne/utils/mixin.py:447]: convert to f-strings please

File: mne/utils/mixin.py [hunk 5]
```diff
@@ -448,9 +473,55 @@ def tmax(self):
...
+
+        Notes
+        -----
+        %(notes_tmax_included_by_default)s
+        """
+        if tmin is None:
+            tmin = self.tmin
+        elif tmin < self.tmin:
+            warn(f'tmin is not in time interval. tmin is set to '
+                 f'inst.tmin ({self.tmin:g} sec)')
```
Comment (drammock) [inline mne/utils/mixin.py:502]: for these warnings, can we figure out from `self` what the class is, and insert the string `Epochs` or `Evoked` or whatever into the warning, in place of `inst`?

File: mne/utils/mixin.py [hunk 6]
```diff
@@ -501,8 +575,107 @@ def decimate(self, decim, offset=0, verbose=None):
         else:
             self._decim_slice = decim_slice
         self._set_times(self._raw_times[self._decim_slice])
+        self._update_first_last()
+        return self
+
+    # Overridden method signature does not match call...
+    def time_as_index(self, times, use_rounding=False):  # lgtm
```
Comment (drammock) [inline mne/utils/mixin.py:582]: can you comment on this and the below methods, to highlight any changed lines (or confirm that it's a straight copy-paste from elsewhere)?

**Review by drammock — COMMENTED** [review]
this looks like a nice refactoring. LMK when it's ready for another look.

**Comment by alexrockhill** [comment]
> > That looks pretty sensible to me for the multiple return type https://output.circle-artifacts.com/output/job/76871b03-098a-4353-9ac0-402f230ccd3b/artifacts/0/dev/generated/mne.Evoked.html?highlight=evoked#mne.Evoked.decimate, definitely an improvement over `main`.
> 
> I'm not sure about the glossary entry for "time-series object".  Try `git grep "inst :"` and you'll see lots of examples of how we handle this situation elsewhere. We're not perfectly consistent in how we do it, but let's not make that worse by adding a new variant.

I looked but I think this is a new case for mixins where you don't have multiple input types accepted (because it's a method) and ideally you would just put the correct output type as the return but there's no way to know the class from the method docstring so I think this is the most elegant solution rather than having a huge list or for me even more confusing evoked methods that say they return epochs.

File: mne/tests/test_evoked.py (hunk 2 above)
Comment (alexrockhill) [reply mne/tests/test_evoked.py:84]: It doesn't work with the immediately following tests with a simpler `[::decim]` because that doesn't account for the API change

File: mne/time_frequency/tests/test_tfr.py (hunk 3 above)
Comment (alexrockhill) [reply mne/time_frequency/tests/test_tfr.py:718]: It warned that `-0.1` wasn't in the time range now, which I just wanted to avoid writing a warning catch

File: mne/utils/mixin.py (hunk 4 above)
Comment (alexrockhill) [reply mne/utils/mixin.py:447]: Sure, that was just moving it though

File: mne/utils/mixin.py (hunk 5 above)
Comment (alexrockhill) [reply mne/utils/mixin.py:502]: Sure

File: mne/utils/mixin.py (hunk 6 above)
Comment (alexrockhill) [reply mne/utils/mixin.py:582]: It is a straight copy-paste from `mne.io.base.py`

**Comment by alexrockhill** [comment]
Ok, I touched the failing examples and everything is green, should be good to go @drammock!

File: mne/time_frequency/tests/test_tfr.py (hunk 3 above)
Comment (drammock) [reply mne/time_frequency/tests/test_tfr.py:718]: ok, it's fine to avoid catching the warning, but the test should test something *sensible*: cropping to a single sample is rarely done.

File: mne/tests/test_evoked.py (hunk 2 above)
Comment (drammock) [reply mne/tests/test_evoked.py:84]: I don't follow. The API change is that decim now shifts which samples it keeps in order to ensure that it keeps the sample at t=0.  That shouldn't mean that there are *no negative times*.

Comment (alexrockhill) [reply mne/tests/test_evoked.py:84]: If you have negative times then checking that the decimation worked properly is super complicated and done in the tests for epochs so better not to do that here.

File: mne/time_frequency/tests/test_tfr.py (hunk 3 above)
Comment (alexrockhill) [reply mne/time_frequency/tests/test_tfr.py:718]: The times are `[0.1, 0.2, 0.3]` and tmax is `0.1` so there's no other options... it's not a very realistic test

File: mne/tests/test_evoked.py (hunk 2 above)
Comment (drammock) [reply mne/tests/test_evoked.py:84]: ah, OK, I see now. The lines 88 and 89 were hidden from me in the GitHub PR review UI.  If those lines are the issue, then it ought to be possible to revert the change to this line, and edit the starting sample (e.g., the `1` in `1::decim`) in lines 88-89 right?

Comment (alexrockhill) [reply mne/tests/test_evoked.py:84]: It's not that simple, see the tests in `test_epochs`. You have to compensate for the shift such that 0 is in the times.

Comment (drammock) [reply mne/tests/test_evoked.py:84]: it *is* that simple. lines 84-90:
```py
    evoked = EvokedArray(data, info, tmin=-1)  # back the way it started
    evoked_dec = evoked.copy().decimate(decim)
    evoked_dec_2 = evoked.copy().decimate(decim, offset=1)
    evoked_dec_3 = evoked.decimate(dec_1).decimate(dec_2)
    assert_array_equal(evoked_dec.data, data[:, 4::decim])    # just change the 
    assert_array_equal(evoked_dec_2.data, data[:, 5::decim])  # starting samp
    assert_array_equal(evoked_dec.data, evoked_dec_3.data)
```

Comment (drammock) [reply mne/tests/test_evoked.py:84]: also, please stop marking conversations as "resolved" just because you disagree. Sometimes it's obvious (like "please make this an f-string") and it's fine, but in more complex cases, whether something is "resolved" is generally best determined by the person who asked the question.

Comment (alexrockhill) [reply mne/tests/test_evoked.py:84]: -1 from me, the numbers `4` and `5` are picked out of nowhere, it's properly tested here:

```
raw, events, picks = _get_data()
events = events[events[:, 2] == 1][:2]
raw.load_data().pick_channels([raw.ch_names[pick] for pick in picks[::30]])
raw.info.normalize_proj()
del picks
sfreq_new = raw.info['sfreq'] / decim
with raw.info._unlock():
    raw.info['lowpass'] = sfreq_new / 12.  # suppress aliasing warnings
pytest.raises(ValueError, epochs.decimate, -1)
pytest.raises(ValueError, epochs.decimate, 2, offset=-1)
pytest.raises(ValueError, epochs.decimate, 2, offset=2)
for this_offset in range(decim):
    epochs = Epochs(raw, events, event_id,
                    tmin=-this_offset / raw.info['sfreq'], tmax=tmax,
                    baseline=None)
    idx_offsets = np.arange(decim) + this_offset
    for offset, idx_offset in zip(np.arange(decim), idx_offsets):
        expected_times = epochs.times[idx_offset::decim]
        expected_data = epochs.get_data()[:, :, idx_offset::decim]
        must_have = offset / float(epochs.info['sfreq'])
        assert (np.isclose(must_have, expected_times).any())
        ep_decim = epochs.copy().decimate(decim, offset)
        assert (np.isclose(must_have, ep_decim.times).any())
        assert_allclose(ep_decim.times, expected_times)
        assert_allclose(ep_decim.get_data(), expected_data)
        assert_equal(ep_decim.info['sfreq'], sfreq_new)

# More complicated cases
epochs = Epochs(raw, events, event_id, tmin, tmax)
expected_data = epochs.get_data()[:, :, ::decim]
expected_times = epochs.times[::decim]
for preload in (True, False):
    # at init
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=decim,
                    preload=preload)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # split between init and afterward
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_1,
                    preload=preload).decimate(dec_2)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_2,
                    preload=preload).decimate(dec_1)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # split between init and afterward, with preload in between
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_1,
                    preload=preload)
    epochs.load_data()
    epochs = epochs.decimate(dec_2)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)
    epochs = Epochs(raw, events, event_id, tmin, tmax, decim=dec_2,
                    preload=preload)
    epochs.load_data()
    epochs = epochs.decimate(dec_1)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # decimate afterward
    epochs = Epochs(raw, events, event_id, tmin, tmax,
                    preload=preload).decimate(decim)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # decimate afterward, with preload in between
    epochs = Epochs(raw, events, event_id, tmin, tmax, preload=preload)
    epochs.load_data()
    epochs.decimate(decim)
    assert_allclose(epochs.get_data(), expected_data)
    assert_allclose(epochs.get_data(), expected_data)
    assert_equal(epochs.info['sfreq'], sfreq_new)
    assert_array_equal(epochs.times, expected_times)

    # test picks when getting data
    picks = [3, 4, 7]
    d1 = epochs.get_data(picks=picks)
    d2 = epochs.get_data()[:, picks]
    assert_array_equal(d1, d2)
```

I meant that I thought I had addressed it but you're free to unresolve, sorry if that offended you.

Comment (drammock) [reply mne/tests/test_evoked.py:84]: it's not plucked from nowhere. `4 == evoked.times.tolist().index(0) - decim` and `5` is because the second one has `offset=1`

Comment (drammock) [reply mne/tests/test_evoked.py:84]:
```py
    evoked = EvokedArray(data, info, tmin=-1)
    zero_idx = evoked.times.tolist().index(0)
    evoked_dec = evoked.copy().decimate(decim)
    evoked_dec_2 = evoked.copy().decimate(decim, offset=1)
    evoked_dec_3 = evoked.copy().decimate(dec_1).decimate(dec_2)
    start_samp = zero_idx - decim
    assert_array_equal(evoked_dec.data, data[:, start_samp::decim])
    # this has +1 because offset=1 when decimating ↓↓↓↓↓↓↓↓↓↓↓↓↓↓
    assert_array_equal(evoked_dec_2.data, data[:, (start_samp + 1)::decim])
```
you can change n_times on line 75 to whatever you like and this will still pass

**Comment by alexrockhill** [comment]
Ok to merge @drammock?

File: mne/utils/mixin.py [hunk 7]
```diff
@@ -501,8 +575,107 @@ def decimate(self, decim, offset=0, verbose=None):
...
+
+    def _handle_tmin_tmax(self, tmin, tmax):
+        """Convert seconds to index into data.
+
+        Parameters
+        ----------
+        tmin : int | float | None
+            Start time of data to get in seconds.
+        tmax : int | float | None
+            End time of data to get in seconds.
```
Comment (mscheltienne) [inline mne/utils/mixin.py:616]: Should this support an `include_tmax` argument similar to `crop`? 
Related to #10372

File: doc/conf.py [hunk 8]
```diff
@@ -237,6 +237,7 @@
     'Transform': 'mne.transforms.Transform',
     'Coregistration': 'mne.coreg.Coregistration',
     'Figure3D': 'mne.viz.Figure3D',
+    'MNE-object': ':term:`MNE time-series object`',
```
Comment (larsoner) [inline doc/conf.py:240]: I'd rather not do this in this PR. In most places we just do `instance of Raw, Epochs, or Evoked`. It's not optimal and in some cases not even complete (e.g., can be AverageTFR, STC, etc. instead) but it's common and people are used to it. If we want to go this route it should be in a separate PR where we change a ton of these at once. But for now I'd just make this `instance of Raw, Epochs, Evoked, AverageTFR, or SourceEstimate` or so.

But even separate-PR aside, I'm not totally convinced we need to or want to create this new category just for the purposes of making our `Returns` section better. For example, to me the best solution would be to somehow magically infer the class name being documented but I don't think this is possible based on how autodoc works. To me the simplest solution is to do something like `instance of original type` or something more compact that has this same meaning to say "you get back the same instance mostly for chaining purposes".

We should discuss this separately in a new issue. @alexrockhill if you're motivated by this problem, could you try making a new issue with the points that people have made about this already, then we can continue the discussion there?

**Review by larsoner — CHANGES_REQUESTED** [review]

File: mne/utils/mixin.py (hunk 7 above)
Comment (larsoner) [reply mne/utils/mixin.py:616]: Perhaps eventually yes... I'll just add a note to #10372 that we should tackle it at the same time as that ideally

File: doc/conf.py (hunk 8 above)
Comment (alexrockhill) [reply doc/conf.py:240]: +1 for instance of original type, it is a tricky corner case for autodoc. Sounds good, I'll remove.

Comment (larsoner) [reply doc/conf.py:240]: ... but for now in this PR please just use a list and open an issue where we discuss separately what the correct verbiage/linking should be for this "original type" stuff would be

**Review by larsoner — APPROVED** [review]

**Comment by larsoner** [comment]
Thanks @alexrockhill !
